File: src/main/java/com/mattiaswikstrom/hyper/Trig.java
package com.mattiaswikstrom.hyper;

import java.util.*;
import java.io.Serializable;

public final class Trig implements Serializable
{
    private double k;
    private double kinv;

    Trig(double k)
    {
        this.k = k;
        this.kinv = 1/k;
    }

    public double getCurvature()
    {
        return k;
    }

    public static double sin(Angle a)
    {
        return Math.sin(a.a);
    }

    public static double cos(Angle a)
    {
        return Math.cos(a.a);
    }

    public static Angle asin(double d)
    {
        return new Angle(Math.asin(d));
    }

    public static Angle acos(double d)
    {
        return new Angle(Math.acos(d));
    }

    public double sigma(Length l)
    {
        if (k == 0)
            return l.l;
        else if (k < 0)
            return Math2.sinh(l.l*k);
        else 
            return Math.sin(l.l*k);
    }

    public Length sigmainv(double d)
    {
        if (k == 0)
            return new Length(d);
        else if (k < 0)
            return new Length(kinv * Math2.asinh(d));
        else
            return new Length(kinv * Math.asin(d));
    }

    public double tau(Length l)
    {
        if (k == 0)
            return 1;
        else if (k < 0)
            return Math2.cosh(l.l*k);
        else 
            return Math.cos(l.l*k);
    }

    public Length tauinv(double d)
    {
        if (k == 0)
            return new Length(0);
        else if (k < 0)
            return new Length(kinv * Math2.acosh(d));
        else 
            return new Length(kinv * Math.acos(d));
    }

    public static double dprod(Length l1, Length l2)
    {
        return 2 * l1.l * l2.l;
    }

    public static double square(Length l)
    {
        return l.l * l.l;
    }

    public Angle angle_from_sss(Length s1, Length s2, Length opp)
    {
        if (k == 0)
            return acos((square(s1)+square(s2)-square(opp)) / dprod(s1, s2));
        else
            return acos((tau(s1)*tau(s2)-tau(opp)) / (sigma(s1)*sigma(s2)));
    }

    public Angle angle_from_aas(Angle a1, Angle a2, Length s)
    {
        if (k == 0)
            return new Angle(Math.PI - a1.a - a2.a);
        else
            return acos(sin(a1)*sin(a2)*tau(s) - cos(a1)*cos(a2));
    }

    Length side_from_aas(Angle a1, Angle a2, Length s)
    {
        return sigmainv(sin(a1)*sigma(s)/sin(a2));
    }

    public Angle angle_from_sas(Length s1, Angle a, Length s2)
    {
        return asin(sigma(s1)*sin(a)/sigma(s2));
    }

    public Length side_from_sas(Length s1, Angle a, Length s2)
    {
        return tauinv(tau(s1)*tau(s2)+sigma(s1)*sigma(s2)*cos(a));
    }

    public Angle angle_from_ssr(Length s1, Length s2)
    {
        return asin(sigma(s2)/sigma(s1));
    }

    public Length adj_from_asr(Angle a, Length s)
    {
        double ret;
        Angle a2;

        if (k == 0)
        {
            a2 = acos(sin(a));
            return adj_from_sra(s, a2); // Math.sin(a2) * Math2.sinh(s) / Math.sin(a);
            //return new Length(s.l / Math.cos(a.a));
        }
        else
        {
            a2 = acos(tau(s)*sin(a));
            return adj_from_sra(s, a2); // Math.sin(a2) * Math2.sinh(s) / Math.sin(a);
        }
    }


    public Length adj_from_sra(Length s, Angle a)
    {
        Angle a2;

        if (k == 0)
                return new Length(s.l / Math.sin(a.a));
        else
        {
                a2 = asin(cos(a) / tau(s));
                return sigmainv(sigma(s) / sin(a));
        }  
    }

    public Length opp_from_sra(Length s, Angle a)
    {
        return sigmainv(sigma(s)/sin(a));
    }

    public Length distance(Length s1, Length s2, Angle a)
    {
        if (k==0)
            return new Length(Math.sqrt(square(s1)+square(s2)-dprod(s1, s2)*cos(a)));
        else
            return tauinv(tau(s1)*tau(s2)
                -sigma(s1)*sigma(s2)*cos(a));
    }

    // Solves a*x*x+b*x+c=0 for x
    private List<Double> solveQuadratic(double a, double b, double c)
    {
        double p = b/a;
        double q = -c/a;
        // New equation: x*x+p*x=q

        double d = q + (p/2)*(p/2);
        List<Double> ret = new ArrayList<Double>(2);

        if (d < 0)
            ;
        else if (d == 0)
        {
            ret.add(Double.valueOf(-p/2));
        }
        else
        {
            double dr = Math.sqrt(d);
            ret.add(Double.valueOf(-p/2 + dr));
            ret.add(Double.valueOf(-p/2 - dr));
        }

        return ret;
    }

    // Solves a * sqrt(1-kappa*x*x)+b*x+c=0 for x
    private List<Double> solveRootEqn(double a, double b, double c, double kappa)
    {
        double a2 = a*a;
        return solveQuadratic(b*b/a2+kappa, 2*b*c/a2, c*c/a2-1);
    }

    public List<Length> side_from_ssa(Length s1, Length s2, Angle a)
    {
        List<Double> s;
        if (k == 0)
        {
            // s1*s1=s2*s2+s3*s3-2*s2*s3*cos(a)
            s = solveQuadratic(1, -2*s2.l*cos(a), square(s2)-square(s1));
        }
        else
        {
            double kappa = k > 0 ? 1 : -1;
            // tau(s1)=tau(s2)tau(s3)+kappa * sigma(s2)sigma(s3)*cos(a)
            // tau(s1)=tau(s2)sqrt(1-kappa*tau(s3)*tau(s3))+kappa*sigma(s2)sigma(s3)*cos(a)
            s = solveRootEqn(tau(s2), kappa*sigma(s2)*cos(a), -tau(s1), kappa);
        }

        List<Length> ret = new ArrayList<Length>(2);

        for (Iterator i = s.iterator(); i.hasNext(); )
        {
            Double d = (Double) i.next();
            ret.add(sigmainv(d.doubleValue()));
        }

        return ret;
    }
}

File: hyper-web/src/Trig.ts
import { Length  } from "./Length.js";
import { Angle  } from "./Angle.js";

export class Trig {
    private k: number;
    private kinv: number;

    constructor(k: number) {
        this.k = k;
        this.kinv = 1 / k;
    }

    getCurvature(): number {
        return this.k;
    }

    static sin(a: Angle): number {
        return Math.sin(a.a);
    }

    static cos(a: Angle): number {
        return Math.cos(a.a);
    }

    static asin(d: number): Angle {
        return new Angle(Math.asin(d));
    }

    static acos(d: number): Angle {
        return new Angle(Math.acos(d));
    }

    /*
        This function appears in the Law of Sines in non-Euclidean geometry:
            sin(α)/σ(a) = sin(β)/σ(b) = sin(γ)/σ(c)
    */
    sigma(l: Length): number {
        if (this.k === 0) return l.l;
        else if (this.k < 0) return Math.sinh(l.l * this.k);
        else return Math.sin(l.l * this.k);
    }

    sigmainv(d: number): Length {
        if (this.k === 0) return new Length(d);
        else if (this.k < 0) return new Length(this.kinv * Math.asinh(d));
        else return new Length(this.kinv * Math.asin(d));
    }

    /*
        This function appears in the Law of Cosines in non-Euclidean geometry:
            τ(c)=τ(a)⋅τ(b)+σ(a)⋅σ(b)⋅cos(γ)
    */
    tau(l: Length): number {
        if (this.k === 0) return 1;
        else if (this.k < 0) return Math.cosh(l.l * this.k);
        else return Math.cos(l.l * this.k);
    }

    tauinv(d: number): Length {
        if (this.k === 0) return new Length(0);
        else if (this.k < 0) return new Length(this.kinv * Math.acosh(d));
        else return new Length(this.kinv * Math.acos(d));
    }

    static dprod(l1: Length, l2: Length): number {
        return 2 * l1.l * l2.l;
    }

    static square(l: Length): number {
        return l.l * l.l;
    }

    angle_from_sss(s1: Length, s2: Length, opp: Length): Angle {
        if (this.k === 0)
            return Trig.acos((Trig.square(s1) + Trig.square(s2) - Trig.square(opp)) / Trig.dprod(s1, s2));
        else
            return Trig.acos((this.tau(s1) * this.tau(s2) - this.tau(opp)) / (this.sigma(s1) * this.sigma(s2)));
    }

    angle_from_aas(a1: Angle, a2: Angle, s: Length): Angle {
        if (this.k === 0)
            return new Angle(Math.PI - a1.a - a2.a);
        else
            return Trig.acos(Trig.sin(a1) * Trig.sin(a2) * this.tau(s) - Trig.cos(a1) * Trig.cos(a2));
    }

    side_from_aas(a1: Angle, a2: Angle, s: Length): Length {
        return this.sigmainv(Trig.sin(a1) * this.sigma(s) / Trig.sin(a2));
    }

    angle_from_sas(s1: Length, a: Angle, s2: Length): Angle {
        return Trig.asin(this.sigma(s1) * Trig.sin(a) / this.sigma(s2));
    }

    side_from_sas(s1: Length, a: Angle, s2: Length): Length {
        if (this.k === 0) {
            // Euclidean case: Use the Law of Cosines
            const cSquared = Trig.square(s1) + Trig.square(s2) - 2 * s1.l * s2.l * Trig.cos(a);
            return new Length(Math.sqrt(cSquared));
        } else {
            // Non-Euclidean case: Use the generalized formula
            return this.tauinv(this.tau(s1) * this.tau(s2) + this.sigma(s1) * this.sigma(s2) * Trig.cos(a));
        }
    }

    angle_from_ssr(s1: Length, s2: Length): Angle {
        return Trig.asin(this.sigma(s2) / this.sigma(s1));
    }

    adj_from_asr(a: Angle, s: Length): Length {
        let a2: Angle;
        if (this.k === 0) {
            a2 = Trig.acos(Trig.sin(a));
            return this.adj_from_sra(s, a2);
        } else {
            a2 = Trig.acos(this.tau(s) * Trig.sin(a));
            return this.adj_from_sra(s, a2);
        }
    }

    adj_from_sra(s: Length, a: Angle): Length {
        if (this.k === 0)
            return new Length(s.l / Trig.sin(a));
        else {
            //let a2: Angle = Trig.asin(Trig.cos(a) / this.tau(s));
            return this.sigmainv(this.sigma(s) / Trig.sin(a));
        }
    }

    opp_from_sra(s: Length, a: Angle): Length {
        return this.sigmainv(this.sigma(s) / Trig.sin(a));
    }

    distance(s1: Length, s2: Length, a: Angle): Length {
        if (this.k === 0)
            return new Length(Math.sqrt(Trig.square(s1) + Trig.square(s2) - Trig.dprod(s1, s2) * Trig.cos(a)));
        else
            return this.tauinv(this.tau(s1) * this.tau(s2) - this.sigma(s1) * this.sigma(s2) * Trig.cos(a));
    }

    private solveQuadratic(a: number, b: number, c: number): number[] {
        const p = b / a;
        const q = -c / a;

        const d = q + (p / 2) * (p / 2);
        let ret: number[] = [];

        if (d >= 0) {
            const dr = Math.sqrt(d);
            ret.push(-p / 2 + dr);
            ret.push(-p / 2 - dr);
        }

        return ret;
    }

    private solveRootEqn(a: number, b: number, c: number, kappa: number): number[] {
        const a2 = a * a;
        return this.solveQuadratic(b * b / a2 + kappa, 2 * b * c / a2, c * c / a2 - 1);
    }

    side_from_ssa(s1: Length, s2: Length, a: Angle): Length[] {
        let s: number[];
        if (this.k === 0) {
            s = this.solveQuadratic(1, -2 * s2.l * Trig.cos(a), Trig.square(s2) - Trig.square(s1));
        } else {
            const kappa = this.k > 0 ? 1 : -1;
            s = this.solveRootEqn(this.tau(s2), kappa * this.sigma(s2) * Trig.cos(a), -this.tau(s1), kappa);
        }

        return s.map(d => this.sigmainv(d));
    }
}